prow/spyglass/gcsartifact.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spyglass
+
+import (
+	"compress/gzip"
+	"context"
+	"io/ioutil"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/spyglass/viewers"
+)
+
+// GCSArtifact represents some output of a prow job stored in GCS
+type GCSArtifact struct {
+	// The handle of the object in GCS
+	handle *storage.ObjectHandle
+
+	// The direct link to the Artifact, can be used for read operations
+	link string
+
+	// The path of the Artifact within the job
+	path string
+
+	// Max size to read before failing
+	sizeLimit int64
+}
+
+// NewGCSArtifact returns a new GCSArtifact with a given handle, canonical link, and path within the job
+func NewGCSArtifact(handle *storage.ObjectHandle, link string, path string, sizeLimit int64) *GCSArtifact {
+	return &GCSArtifact{
+		handle:    handle,
+		link:      link,
+		path:      path,
+		sizeLimit: sizeLimit,
+	}
+}
+
+func fieldsFor(a *GCSArtifact) logrus.Fields {
+	return logrus.Fields{
+		"artifact": a.path,
+	}
+}
+
+// Size returns the size of the artifact in GCS
+func (a *GCSArtifact) Size() (int64, error) {
+	attrs, err := a.handle.Attrs(context.Background())
+	if err != nil {
+		return 0, err
+	}
+	return attrs.Size, nil
+}
+
+// JobPath gets the GCS path of the artifact within the current job
+func (a *GCSArtifact) JobPath() string {
+	return a.path
+}
+
+// CanonicalLink gets the GCS web address of the artifact
+func (a *GCSArtifact) CanonicalLink() string {
+	return a.link
+}
+
+// ReadAt reads len(p) bytes from a file in GCS at offset off
+func (a *GCSArtifact) ReadAt(p []byte, off int64) (n int, err error) {
+	var ctx context.Context
```

Comment:
Seems like context should be part of GCSArtifact, which can set this to `context.Background()` initially and allow user to override as appropriate. Then just pass `a.context` whever needed